### ISSUE TITLE
New Action Creators Extension

### ISSUE THREAD

**Issue by sprjr** [jmeas/redux-resource#186]
I'm not sure how I feel about that title. But I didn't want to bikeshed it too much because what I really want to bikeshed is _this stuff down there_. 

⬇️ 
---

Currently, if you leverage methods from the `resourceful-action-creators`, it does two really neat things for you. For me, those things are in this order of value:

1. Help enforce resourceful-normalized API responses (either naturally or with `transformData` so that they can be properly dispatched into the `resourceful-redux` reducers. The major value add for me here comes from being able to reason about a given resource slice (e.g. `books`) in my React applications, especially when coupled with `getStatus()` to enhance resources in UIs with spinners, retries, cancels, etc.
2. Automatically manage the XHR requests for you with an easy-to-proxy API to the underlying `xhr` module for making resource related requests.

There's obviously more to it than that, but that is where I typically reach for this module when building. One downside to this however, is that if you ever want to take back control of that second portion (e.g. by using something like WebSockets, observables/streaming responses, or even just your _own_ specific XHR edge cases or preferred libraries), you lose out on the benefits of the first portion.

Users are definitely still able to make their own resource requests, and then invoke `dispatch(...)` on their own. The only tricky part is, that they need to really understand how the reducers expect action objects to be formed. It's not impossible of course, it looks something like this:

```js
dispatch({
  ...options,
  type: actionTypes[`${crudType}_RESOURCES_SUCCEEDED`],
  statusCode,
  resources,
  res
});
```

You can see more about what the `options` object is [here](https://resourceful-redux.js.org/docs/extras/resourceful-action-creators.html#crudaction-options-callback-) as I think that's probably a good API to try to model from when creating what I think _could_ be called: *resourceful-dispatch-creators* that could possible be it's own module for the ecosystem.

Imagine an action creator that did *not* want to loose control of it's own XHR (say it could manage cancel-ability, duplicate request logic, retries, etc). Right now, you'd have to make sure that you dispatch the correct shapes for the correct reducers. But many of those options _could_ be provided for you by default - then you'd really only need to worry about passing the resources or the error from an app developer standpoint.

Here's an example of how I think it could work, compared with the example [here](https://resourceful-redux.js.org/docs/extras/resourceful-action-creators.html#tips)

```js
import { dispatchCreators } from 'resourceful-dispatch-creators';

function readManyBooks(pageNumber) {
  // instead of configuring `readResources` here, we're providing the same info,
  // but this will be used to create our final `dispatch(...)` that our reducers pick up
  const dispatchers = dispatchCreators({
    crudAction: 'read',
    resourceName: 'books',
    label: 'homePageBooks',
    mergeLabelIds: false,
  });

  // we make our own preferred XHR request (or open a stream, etc) here
  // and then we use the same `.{status}` method that we'd expect to see results
  // from in `getStatus()` calls in our UIs

  dispatchers.pending();

  return fetch('/books')
    .then(res => dispatchers.succeeded(res))
    .catch(err => dispatchers.failed(err));
}
```

Not 100% sure of the naming, or API, or if this would be useful for folks. But thought of it today while talking with @jmeas about #185 and how that may path us toward building out something really robust like `ember-concurrency` - which may still be useful for the fully formed action creators library, but would further lock folks into our XHR decisions and patterns.

This could also exist and be useful for folks who want more control, simple APIs, and flexibility to do things how they really need to in their edge cases.

Thoughts? /cc @vinodkl @marlonpp @tbranyen

**Comment by jmeas** [comment]
This seems like a good idea to me. I can imagine refactoring action-creators to use this lower-level lib.

My only bit of feedback is that we shouldn’t use the word “dispatch” unless they’re bound, and they probably shouldn’t be bound. So maybe something like `crudActionHelpers` or something (to distinguish them from the action-creators)

:+1:

**Comment by tbranyen** [comment]
Updated your comment to fix some syntax errors @sprjr, I like the API :+1:

**Comment by sprjr** [comment]
@jmeas agreed, I am not sold on the same either. The _weird_ thing is, these are really just action creators with a resourceful API over them. You give them some data, and they return an object that you'd probably dispatch yourself now that I think about it. So I was kinda confused because the current action creators are like, super-duper ones that do thunking for you.

@tbranyen thanks! I'm terrible at pseudo. Well, and real code too.

**Comment by sprjr** [comment]
Also just noticing that my code doesn't actually work here at all. It would need to be more like:

```js

export action = (dispatch) => (opts) => {
  return fetch('/foo')
    .then(res => dispatch(dispatchers.succeeded(res));
}
```

**Comment by jmeas** [comment]
The existing `resourceful-action-creators` lib has been renamed and published as `resourceful-xhr` to allow the lib described in this issue to use the name `resourceful-action-creators`.

When we publish this lib, we'll just need to make sure that the first release is`2.0.0`.

:v:

**Comment by sprjr** [comment]
Bike shed gist of the API is [here](https://gist.github.com/sprjr/3985d85de334e98852f8c1de332afb17). I will try to get an early PR out so we can try an `-alpha.1` release in a few apps/plugins and see how that works In The Real World™

**Comment by jmeas** [comment]
Resolved in #250
